test/common/buffer/owned_impl_test.cc

```diff
@@ -0,0 +1,46 @@
+#include "common/buffer/buffer_impl.h"
+
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace Buffer {
+namespace {
+
+class OwnedImplTest : public testing::Test {
+public:
+  OwnedImplTest() {}
+
+  bool release_callback_called_ = false;
+};
+
+TEST_F(OwnedImplTest, AddBufferFragmentNoCleanup) {
+  char input[] = "hello world";
+  BufferFragmentImpl frag(input, 11, nullptr);
+  Buffer::OwnedImpl buffer;
+  buffer.addBufferFragment(&frag);
+  EXPECT_EQ(11, buffer.length());
+
+  buffer.drain(11);
+  EXPECT_EQ(0, buffer.length());
+}
+
+TEST_F(OwnedImplTest, addBufferFragmentWithCleanup) {
+  char input[] = "hello world";
+  BufferFragmentImpl frag(input, 11,
+                          [this](const void*, size_t) { release_callback_called_ = true; });
```

Comment:
This is fine. Another way to structure it though would be to have separate tests for `BufferFragmentImpl`, which are only about its API, and another for `addBufferFragment`, which use a mock implementation of `BufferFragment`. That would be logically a little cleaner, but /shrug.

Reply:
Hmm, probably not a lot of added value in doing that. If BufferFragmentImpl ever becomes more, like ref counting perhaps, would be good to test separately. For now I think this should be fine.